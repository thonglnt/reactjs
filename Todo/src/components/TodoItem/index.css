.todo-item {
  list-style: none;
  display: flex;
  align-items: center;
  position: relative;
  border: 1px solid var(--light-gray);
  padding-left: 20px;
}

.round {
  position: relative;
}

.round label {
  background-color: var(--white);
  border: 1px solid var(--chinese-silver);
  border-radius: 50%;
  cursor: pointer;
  height: 20px;
  left: 0;
  position: absolute;
  top: 0;
  width: 20px;
}

.round label:after {
  border: 2px solid var(--mantis);
  border-top: none;
  border-right: none;
  content: '';
  height: 6px;
  left: 5px;
  opacity: 0;
  position: absolute;
  top: 5px;
  transform: rotate(-45deg);
  width: 9px;
}

.round input[type='checkbox'] {
  visibility: hidden;
}

.round input[type='checkbox']:checked + label {
  border-color: var(--mantis);
}

.round input[type='checkbox']:checked + label:after {
  opacity: 1;
}

.todo-task {
  padding-left: 20px;
}
