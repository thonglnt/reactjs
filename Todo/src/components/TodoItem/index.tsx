// Assets
import { iconDelete } from '@assets';

// Components
import Spinner from '@components/Spinner';

export type TodoItemProps = {
  id: string;
  task: string;
  status: boolean;
  onRemoveTodo: (id: string) => void;
  onToggleTodo: (id: string) => void;
  isRemoving?: boolean;
};

const TodoItem = ({
  id,
  task,
  status = false,
  onRemoveTodo,
  onToggleTodo,
  isRemoving = false,
}: TodoItemProps) => {
  const handleRemoveTodo = () => {
    onRemoveTodo(id);
  };

  const handleToggleTodo = () => {
    onToggleTodo(id);
  };

  return (
    <li className='todo-item'>
      <div className='round'>
        <input checked={status} onChange={handleToggleTodo} type='checkbox' />
        <label data-testid='check-done-todo' onClick={handleToggleTodo} htmlFor='checkbox' />
      </div>
      <p className='todo-task'>{status ? <s>{task}</s> : task}</p>
      {isRemoving ? (
        <Spinner className='icon' />
      ) : (
        <img
          src={iconDelete}
          data-testid='delete-todo'
          alt='delete'
          className='icon'
          onClick={handleRemoveTodo}
        />
      )}
    </li>
  );
};

export default TodoItem;
