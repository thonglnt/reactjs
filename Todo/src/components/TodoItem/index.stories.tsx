// Lib
import { Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';

// Components
import TodoItem, { TodoItemProps } from '@components/TodoItem';

export default {
  title: 'TodoItem',
  component: TodoItem,
};

export const Sample: Story<TodoItemProps> = () => (
  <TodoItem
    id='1'
    task='todo'
    status={false}
    onRemoveTodo={action('onRemoveTodo')}
    onToggleTodo={action('onToggleTodo')}
  />
);
