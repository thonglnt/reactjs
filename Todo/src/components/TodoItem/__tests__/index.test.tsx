// Lib
import { render, screen, fireEvent } from '@testing-library/react';

// Components
import TodoItem, { TodoItemProps } from '@components/TodoItem';

describe('Test TodoItem component', () => {
  const initTodoItem: TodoItemProps = {
    id: '1',
    task: 'task 1',
    status: false,
    onRemoveTodo: jest.fn(),
    onToggleTodo: jest.fn(),
  };

  test('It should render correctly with initial Props', () => {
    const { container } = render(<TodoItem {...initTodoItem} />);

    expect(container).toMatchSnapshot();
  });

  test('It should call remove todo', () => {
    render(<TodoItem {...initTodoItem} />);
    fireEvent.click(screen.getByTestId('delete-todo'));

    expect(initTodoItem.onRemoveTodo).toHaveBeenCalledTimes(1);
  });

  test('It should call on check done todo', () => {
    render(<TodoItem {...initTodoItem} />);
    fireEvent.click(screen.getByTestId('check-done-todo'));

    expect(initTodoItem.onRemoveTodo).toHaveBeenCalledTimes(1);
  });
});
