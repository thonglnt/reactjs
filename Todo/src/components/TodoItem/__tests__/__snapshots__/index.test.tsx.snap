// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoItem component It should render correctly with initial Props 1`] = `
<div>
  <li
    class="todo-item"
  >
    <div
      class="round"
    >
      <input
        type="checkbox"
      />
      <label
        data-testid="check-done-todo"
        for="checkbox"
      />
    </div>
    <p
      class="todo-task"
    >
      task 1
    </p>
    <img
      alt="delete"
      class="icon"
      data-testid="delete-todo"
      src="IMAGE_MOCK"
    />
  </li>
</div>
`;
